### PR TITLE
Client for external metrics

### PR DESCRIPTION
**Pull request by MaciekPytel**
Add client for External Metrics API introduced in https://github.com/kubernetes/kubernetes/pull/60079.

This PR is just client all the API changes are included from https://github.com/kubernetes/kubernetes/pull/60079.

### CODE REVIEW

**Comment by MaciekPytel** [comment]
/hold
This depends on https://github.com/kubernetes/kubernetes/pull/60079 and must be hold until it's merged.

File: staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go
Comment (DirectXMan12) [inline staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: Any reason this isn't just a `list` operation?  It seems like that actually makes some amount of sense (we're "listing" metric values)...

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
comments inline

File: staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go
Comment (MaciekPytel) [reply staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: You're right, it makes more sense. Fixed.

**Comment by DirectXMan12** [comment]
Since this is list now, this should just all be autogenerated, like for the resource metrics API.  It'll save us headaches later.

**Comment by MaciekPytel** [comment]
@DirectXMan12 updated as per our discussion offline, can you take a look?

For reference - we discussed offline and it looks like we won't be able to use autogenerated client after all.

**Comment by MaciekPytel** [comment]
/unhold

The PR this depended on has been merged and I rebased this PR.

File: staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go
Comment (liggitt) [inline staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: this, combined with the lack of any name validation in the API, is pretty concerning. dropping unvalidated names into a URL builder is a good way to introduce path traversal attacks (e.g. "my metric name is `..%2F..%2F..%2F..%2F..%2F..%2Fapi%2Fv1%2Fnamespaces%2Fkube-system%2Fsecrets`")

cc @kubernetes/api-reviewers

File: staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go
Comment (liggitt) [inline staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: requiring client-side fixup for the field that is supposed to uniquely identify the object looks very strange

File: staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go
Comment (liggitt) [reply staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: FYI, `Resource()` will reject any name that does not pass `rest.IsValidPathSegmentName`, which disallows names that contain `/`, `%`, or are exactly `.`, `..`

the API field driving this should pass the same validation

File: staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go
Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: agreed.  I very much dislike this escape.

Comment (MaciekPytel) [reply staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go]: removed

**Review by DirectXMan12 — APPROVED** [review]

**Pull request merged**
